flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamOperator.java

```diff
@@ -43,7 +43,7 @@
  * @param <OUT> The output type of the operator
  */
 @PublicEvolving
-public interface StreamOperator<OUT> extends CheckpointListener, KeyContext, Serializable {
+public interface StreamOperator<OUT> extends InternalCheckpointListener, KeyContext, Serializable {
```

Comment:
It seems wrong that a `@PublicEvolving` interface extends `@Internal` interface.
The root cause I think in the `StreamOperator` hierarchy, but that's a long story.

Semantically, I think operators should only be aware of checkpoint completion (to emit data for example); their backends should in addition know about the subsumption.

I see the following solutions to notify state backends about the subsumption
1. Modify `CheckpointListener` interface instead of introducing `InternalCheckpointListener` as you proposed initially
2. In `StreamOperatorWrapper`, get  `KeyedStateBackend` from the operator (if it's `AbstractStreamOperator`) and notify it directly (if it's `InternalCheckpointListener`)
3. Same, but via `StreamOperatorStateHandler`

WDYT?

cc: @pnowojski

Reply:
> It seems wrong to me that a @PublicEvolving interface extends @Internal interface.

+1 

What do you @rkhachatryan mean by:
> 1. Modify CheckpointListener interface instead of introducing InternalCheckpointListener as you proposed initially

?

Regardless I also think we shouldn't go through the operators if we don't want to make `notifyCheckpointSubsumed()` public.

Reply:
I meant to add a method `checkpointSubsumed()` to the `CheckpointListener` (or `lastSubsumedId` argument to its existing `notifyCheckpointComplete()` method).

Reply:
Do we want to expose `notifyCheckpointSubsumed()` on the DataStream API level in a `@Public` class?

Reply:
@pnowojski I think this needs more discussion here.
My previous idea was to add a default no-op in `CheckpointListener` so that could be public. Roman convinced me that this new interface does not make so much senses for most users, and could be acted as an internal API just like what `InternalKvState` did. 
However, current PR has to change several places to make the logic more clean, which makes me to think whether previous `AbstractStreamOperator` did not welcome such internal semantics here.
The main usage of `notifyCheckpointSubsumed` is for abstract keyed state backend, we can compare the relationship of `AbstractStreamOperator` with `AbstractKeyedStateBackend`. If `AbstractStreamOperator` is open for users to implement and `AbstractKeyedStateBackend` is actually also open for users to implement their own state backend.
If so, we might treat `notifyCheckpointSubsumed` as public also.
The only left questions are can we add a new method in `@Public` class and we need to remember to implement it in necessary places.
cc @rkhachatryan

Reply:
I'm also not sure if it makes sense to expose `notifyCheckpointSubsumed()` to the users. And if we don't have a good motivating example, I would prefer to not expose it. Why clog the public API at best? And at worst this could cause us problems down the road for providing backward compatibility/refactoring. If needed, we can always add this notification later. Removing and maintaining things is harder.